src/shogun/kernel/GaussianARDKernel.cpp

```diff
@@ -122,21 +107,10 @@ SGMatrix<float64_t> CGaussianARDKernel::get_parameter_gradient(
 		{
 			for (index_t k=0; k<num_rhs; k++)
 			{
-				SGVector<float64_t> avec=
-					((CDenseFeatures<float64_t>*) lhs)->get_feature_vector(j);
-				SGVector<float64_t> bvec=
-					((CDenseFeatures<float64_t>*) rhs)->get_feature_vector(k);
-
-				REQUIRE(avec.vlen==bvec.vlen, "Number of right and left hand "
-						"features must be the same\n");
-
-				float64_t result=0;
+				float64_t result=distance(j,k);
```

Comment:
Computing pairwise distances is the same though.

In fact, what about the following:
Rather than having different classes for ARD kernels, why not add an enum to the existing class.
KT_SCALAR, KT_DIAG, KT_FULL
for scalar (same as std), diagonal (your ARD kernel), and full pairwise (a symmetric matrix) weights?
Then all these seperate classes would just go. Let me know what you think?